znnl/utils/matrix_utils.py

```diff
@@ -134,3 +139,51 @@ def calculate_l_pq_norm(matrix: np.ndarray, p: int = 2, q: int = 2):
     inner_sum = np.sum(np.power(matrix, q), axis=-1)
     outer_sum = np.sum(np.power(inner_sum, q / p), axis=-1)
     return np.power(outer_sum, 1 / q)
+
+
+def flatten_rank_4_tensor(tensor: np.ndarray) -> np.ndarray:
+    """
+    Flatten a rank 4 tensor to a rank 2 tensor using a specific reshaping.
+
+    The tensor is assumed to be of shape (n, n, m, m). The reshaping is done by
+    concatenating first with the third and then with the fourth dimension, resulting
+    in a tensor of shape (n * m, n * m).
+
+    Parameters
+    ----------
+    tensor : np.ndarray (shape=(n, n, m, m))
+            Tensor to flatten.
+
+    Returns
+    -------
+    flattened_tensor : np.ndarray (shape=(n * m, n * m))
+            Flattened tensor.
+    """
+
+    assert tensor.shape[0] == tensor.shape[1]
```

Comment:
Using asserts in production code is kind of a controversial issue. I think you have to ask whether or not this will ever happen if the code is called internally only. Further, it might be better to use an if statement and write some log. You can use a try except routine and manually raise an assertion error with a custom warming.

```python
try:
...
except:
    raise ValueError("Matrix of incorrect shape")
```